compiler/rustc_typeck/src/check/fn_ctxt/checks.rs

```diff
@@ -412,20 +346,603 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                    if permutation[i].is_some() {
+                        // We've already decided whether this argument is or is not in a loop
+                        continue;
+                    }
+
+                    let mut stack = vec![];
+                    let mut j = i;
+                    let mut last = i;
+                    let mut is_cycle = true;
+                    loop {
+                        stack.push(j);
+                        // Look for params this one could slot into
+                        let compat: Vec<_> = mat[j]
+                            .iter()
+                            .enumerate()
+                            .filter_map(|(i, c)| {
+                                if matches!(c, Compatibility::Compatible) { Some(i) } else { None }
+                            })
+                            .collect();
+                        if compat.len() != 1 {
+                            // this could go into multiple slots, don't bother exploring both
+                            is_cycle = false;
+                            break;
+                        }
+                        j = compat[0];
+                        if stack.contains(&j) {
+                            last = j;
+                            break;
+                        }
+                    }
+                    if stack.len() <= 2 {
+                        // If we encounter a cycle of 1 or 2 elements, we'll let the
+                        // "satisfy" and "swap" code above handle those
+                        is_cycle = false;
+                    }
+                    // We've built up some chain, some of which might be a cycle
+                    // ex: [1,2,3,4]; last = 2; j = 2;
+                    // So, we want to mark 4, 3, and 2 as part of a permutation
+                    permutation_found = is_cycle;
+                    while let Some(x) = stack.pop() {
+                        if is_cycle {
+                            permutation[x] = Some(Some(j));
+                            j = x;
+                            if j == last {
+                                // From here on out, we're a tail leading into a cycle,
+                                // not the cycle itself
+                                is_cycle = false;
+                            }
+                        } else {
+                            // Some(None) ensures we save time by skipping this argument again
+                            permutation[x] = Some(None);
+                        }
+                    }
+                }
+
+                if permutation_found {
+                    // Map unwrap to remove the first layer of Some
+                    let final_permutation: Vec<Option<usize>> =
+                        permutation.into_iter().map(|x| x.unwrap()).collect();
+                    return Some(Issue::Permutation(final_permutation));
+                }
+                return None;
+            };
```

Comment:
Everything since line 430 could be its own struct and impl, so that we can remove all of this logic out of the diagnostic code and make it clearer what is happening to anyone reading the code as prose.